internal/runtime/catalog/catalog.go

```diff
@@ -0,0 +1,333 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package catalog
+
+import (
+	"fmt"
+	"reflect"
+	goruntime "runtime"
+	"strings"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/naming"
+	"k8s.io/kube-openapi/pkg/common"
+)
+
+// Catalog contains all information about RuntimeHooks defined in Cluster API,
+// including mappings between RuntimeHook functions and the corresponding GroupVersionHooks,
+// metadata and OpenAPIDefinitions.
+type Catalog struct {
+	// scheme is used to access to api-machinery utilities to implement conversions of
+	// request and response types.
+	scheme *runtime.Scheme
+
+	// gvhToType maps a GroupVersionHook to the corresponding RuntimeHook function.
+	gvhToType map[GroupVersionHook]reflect.Type
+
+	// gvhToType maps a RuntimeHook to the corresponding GroupVersionHook.
+	typeToGVH map[reflect.Type]GroupVersionHook
+
+	// gvhToHookDescriptor maps a GroupVersionHook to the corresponding hook descriptor.
+	gvhToHookDescriptor map[GroupVersionHook]hookDescriptor
+
+	// gvToOpenAPIDefinitions maps a GroupVersion to a func
+	// which returns OpenAPI definitions for all request and response types with that
+	// GroupVersion.
+	gvToOpenAPIDefinitions map[schema.GroupVersion]OpenAPIDefinitionsGetter
+
+	// catalogName is the name of this catalog. It is set based on the stack of the New caller.
+	// This is useful for error reporting to indicate the origin of the Catalog.
+	catalogName string
+}
+
+// Hook is a marker interface for a hook.
+// Hooks should be defined as a: func(*RequestType, *ResponseType).
+type Hook interface{}
+
+// hookDescriptor is a data structure which holds
+// all information about a Hook.
+type hookDescriptor struct {
+	metadata *HookMeta
+
+	// request gvk for the Hook.
+	request schema.GroupVersionKind
+	// response gvk for the Hook.
+	response schema.GroupVersionKind
+}
+
+// HookMeta holds metadata for a Hook, which is used to generate
+// the OpenAPI definition for a Hook.
+type HookMeta struct {
+	// Summary of the hook.
+	Summary string
+
+	// Description of the hook.
+	Description string
+
+	// Tags of the hook.
+	Tags []string
+
+	// Deprecated signals if the hook is deprecated.
+	Deprecated bool
+
+	// Singleton signals if the hook can only be implemented once on a
+	// Runtime Extension, e.g. like the DiscoveryHook.
+	Singleton bool
```

Comment:
This field becomes important when we are going to do OpenAPI spec generation. This can be used to distinguish between endpoints that are name specific vs endpoints that are not (singletons).
Example:
The Before Cluster Create hook is not a singleton and the end point will be name specific. Endpoint: `/hooks.runtime.cluster.x-k8s.io/v1alpha1/beforeclusterupgrade/{name}`.
The discovery endpoint is a singleton and the endpoint will be something like: `/hooks.runtime.cluster.x-k8s.io/v1alpha1/discovery`

Reply:
kk, 
I'm not super happy to expose a knob for a what is a single, well-known special case now, but we can eventually re-iterate later (it could be also a simple rename to something that makes it more clear what are the implications of a singleton in this context).
Feel free to resolve the comment

Reply:
I would also try to avoid it, but I wasn't sure how. I didn't want to hard-code `if hook == "Discovery"` but maybe we should just do that. Let's reconsider on the runtime openapi gen PR.